core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -0,0 +1,511 @@
+use std::{ops, time::Instant};
+
+use anyhow::Context;
+use zksync_dal::CoreDal;
+use zksync_multivm::{
+    interface::{TransactionExecutionMetrics, VmExecutionResultAndLogs},
+    utils::{
+        adjust_pubdata_price_for_tx, derive_base_fee_and_gas_per_pubdata, derive_overhead,
+        get_max_batch_gas_limit,
+    },
+    zk_evm_latest::ethereum_types::H256,
+};
+use zksync_system_constants::MAX_L2_TX_GAS_LIMIT;
+use zksync_types::{
+    api::state_override::StateOverride, fee::Fee, fee_model::BatchFeeInput, get_code_key,
+    ExecuteTransactionCommon, PackedEthSignature, ProtocolVersionId, Transaction,
+};
+
+use super::{result::ApiCallResult, SubmitTxError, TxSender};
+use crate::execution_sandbox::{BlockArgs, SandboxAction, VmPermit, SANDBOX_METRICS};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum BinarySearchKind {
+    /// Full binary search.
+    Full,
+    /// Binary search with an optimized initial pivot.
+    #[allow(dead_code)] // FIXME
+    Optimized,
```

Comment:
Do we want to use optimized binary search from the get go, or make it opt-in? Right now, it's only used in tests, hence `allow(dead_code)`. Even without enabling it, we can still collect metrics for the lower bound / initial pivot (see below).